### ISSUE TITLE
Расширить синтаксис break в циклах для выхода из глубокой вложенности

### ISSUE THREAD

**Issue by Neargye** [cpp-ru/ideas#153]
_Перенос предложения:_ голоса +5, -10
_Автор идеи:_ **Игорь Савенков**

Добавить возможность указывать количество циклов из которых нужно выйти

Например:
```cpp
for (int i =0; i < 10; i++)
{
    for (int j = 0; j < 10; j++)
    {
        if (some1)
            break 2;
    }
}
// <- continue from here
```

**Comment by Neargye** [comment]
**al-mission-2016 25 июня 2019, 4:24**
Любопытно, но ненужно. Есть же уже божественное: goto CONTINUE_FROM_HERE;


**Юрий Николаев 25 июня 2019, 12:24**
```cpp
for (int i = 0; i < 10; i++)
{
	switch (i)
	{
		case 1:
		{
			for (int j = 0; j < 10; j++)
			{
				switch (j)
				{
					case 1:
					{
						if (some1)
							break 2;
					}
					break;
				}
			}
			// <- continue from here?
		}
		break;
	}
	// <- or continue from here?
}
// <- or continue from here?
```
Куда же должны попасть в данной ситуации?

Не, перегружать break идея плохая.


**Игорь Савенков 25 июня 2019, 14:06*
Юрий Николаев, я предполагаю, что такой синтаксис break 2 должен работать исключительно для циклов. Т.е. выход будет в самую последнюю точку "continue from here" игнорируя все switch.

Как бы да, получается несколько запутанный код, но надо понимать, что человек пишущий подобную глубину вложения, комбинируя циклы и switch - сам себе злобный буратино.

Альтернатива - новый оператор exitloop 2, например, хотя это overhead для такого частного случая


**Igor Baidiuk 2 июля 2019, 13:34**
Как по мне, если уж делать - то через метки на блоках и labeled break/continue. Получается GOTO без кучи проблем GOTO.


**Григорий Скурко 12 июля 2019, 11:51**
Напрашивается такую логику/пример выделить в функцию и делать return

Если уровней много и нужно между ними скакать - тоже самое - с подобным break тяжело будет код понимать


**NeoProgramming 19 июля 2019, 21:19**
Для этого лучшее решение - именованные блоки. То есть любой блок (if, while, for, switch...) может иметь опциональное имя, которое пишется после круглых скобок и перед телом блока, и которое можно использовать в break.
```cpp
if(x>0) checkX {
  while(i != 100) MyLoop1 {
    switch(c) SelectAction {
      //... 
      break checkX;
      //...
    }
  }
}
```
Но я такое даже не предлагаю - слишком радикальное изменение в языке, никто на это не пойдет.


**valera_ee 3 августа 2019, 11:08**
Подобное сделано в языке Swift, но синтаксис у них для этого напоминает goto в c++

**Comment by Panzerschrek** [comment]
Есть подвижки в эту сторону? Ила всё заглохло и придётся и дальше использовать goto?

**Comment by kov-serg** [comment]
Чем вам goto не устраивает. Или просто религиозная неприязнь. Если так, то можно анонимные функции использовать.
```c++
([&]{
  for (int i =0; i < 10; i++) {
    for (int j = 0; j < 10; j++) {
      if (some1) return;
    }
  }
})();
// <- continue from here
```

**Comment by Panzerschrek** [comment]
`goto` не разрешает перепрыгивать через объявления переменных. Но в случае `goto` для эмуляции `continue` к внешнему циклу могут быть подобные объявления, ибо метка ставится в конце блока. Нормальный `continue` с указанием метки цикла решил бы эту проблему.

Пример с лямбдой выше может работать для замены `break`, но не для `continue`. Ну и в целом это не очень правильно - городить лишнюю сущность (лямбду) для реализации того, что должно бы быть реализовано посредством стандартных операторов передачи потока управления.

**Comment by kov-serg** [comment]
Так вы не перепрыгивайте:
```c++
void example() {
	struct A { const char* name;
		A(const char* name) : name(name) { printf(" +%s",name); } 
		~A() { printf(" -%s",name); } 
	};
	A a0("a0");
	for(int i=0;i<3;i++) {{
		A a1("a1");
		for(int j=0;j<3;j++) {{
			printf(" i=%d j=%d\n",i,j);
			A a2("a2");
			if (i==1 && j==1) goto L1_break;
			if (i==1 && j==0) goto L1_continue;
			if (i==2 && j==1) goto L2_continue;
			if (i==0 && j==2) goto L2_break;
			A a3("a3");
		}L2_continue:{}}L2_break:{}
		A a4("a4");
	}L1_continue:{}}L1_break:{}
}
```

**Comment by Panzerschrek** [comment]
Да, действительно, если обрамить код до метки в `{}`, то всё будет работать. Но так или иначё, всё же было бы неплохо иметь в C++ нативную возможность break/continue для внешнего цикла. Многие другие языки ведь так могут. К тому же можно будет после такого нововведения `goto` сделать deprecated, ну или почти depricated, ибо других применений для него почти что не останется.

**Comment by kov-serg** [comment]
> после такого нововведения goto сделать deprecated

Вот не надо делать не обоснованные выводы. Его и макросы не просто так не убирают. Иначе всё сломается и такой c++ будет никому не нужен.
